File: python/lexscope/audit.py
"""Offline audit verification — identical semantics to src/core/audit.ts.

Verifies a complete AuditPage against an AuditTrust: hash chain, entry
signatures, sequence continuity, signed head, key-rotation admission, and
pinned-head completeness.
"""
from __future__ import annotations

from .canon import jcs
from .crypto import b64u, b64u_decode, domain_hash, sign, verify
from .ids import is_counter, is_hash, is_id

GENESIS_PREV = "0" * 64

AUDIT_FAIL = (
    "AUDIT_HASH_MISMATCH", "AUDIT_INCOMPLETE", "AUDIT_SIGNATURE_INVALID",
    "AUDIT_SEQUENCE_INVALID", "AUDIT_TRUST_INVALID",
)


def entry_hash(body: dict) -> str:
    return domain_hash("LEXSCOPE-AUDIT/1", body)


def _entry_sign_msg(digest_hex: str) -> bytes:
    return b"LEXSCOPE-AUDIT-SIGN/1" + b"\x00" + bytes.fromhex(digest_hex)


def sign_entry(body: dict, priv) -> dict:
    digest = entry_hash(body)
    return {"body": body, "hash": digest, "signature": b64u(sign(priv, _entry_sign_msg(digest)))}


def verify_entry(entry: dict, jwk: dict) -> bool:
    sig = b64u_decode(entry["signature"])
    if sig is None or len(sig) != 64:
        return False
    digest = entry_hash(entry["body"])
    if digest != entry["hash"]:
        return False
    return verify(jwk, _entry_sign_msg(digest), sig)


def sign_head(tenant_id: str, seq: str, h: str, signer_kid: str, priv) -> dict:
    msg = b"LEXSCOPE-HEAD/1" + b"\x00" + jcs(
        {"tenant_id": tenant_id, "seq": seq, "hash": h, "signer_kid": signer_kid}
    )
    return {"seq": seq, "hash": h, "signer_kid": signer_kid, "signature": b64u(sign(priv, msg))}


def verify_head(head: dict, tenant_id: str, jwk: dict) -> bool:
    sig = b64u_decode(head["signature"])
    if sig is None or len(sig) != 64:
        return False
    msg = b"LEXSCOPE-HEAD/1" + b"\x00" + jcs(
        {"tenant_id": tenant_id, "seq": head["seq"], "hash": head["hash"], "signer_kid": head["signer_kid"]}
    )
    return verify(jwk, msg, sig)


def verify_audit(page: dict, trust: dict) -> dict:
    """Returns {"valid": True, "entries": n, "head": hash} or
    {"valid": False, "code": AUDIT_*, "seq": str}."""
    def fail(code, seq):
        return {"valid": False, "code": code, "seq": seq}

    entries = page["entries"]
    key_of = {k["kid"]: k["jwk"] for k in trust["keys"]}

    prev = GENESIS_PREV
    expect_seq = 1
    # offline verification begins with AuditTrust.initial_kid and may advance
    # only at key.rotated events whose new key is in trust.keys
    current_kid = trust["initial_kid"]
    if trust.get("checkpoint") is not None:
        cp = trust["checkpoint"]
        prev = cp["hash"]
        expect_seq = int(cp["body"]["seq"]) + 1
        current_kid = cp["body"]["signer_kid"]

    last = None
    for e in entries:
        body = e["body"]
        if int(body["seq"]) != expect_seq:
            return fail("AUDIT_SEQUENCE_INVALID", body["seq"])
        if body["prev"] != prev:
            return fail("AUDIT_HASH_MISMATCH", body["seq"])
        if body["signer_kid"] != current_kid:
            return fail("AUDIT_TRUST_INVALID", body["seq"])
        if entry_hash(body) != e["hash"]:
            return fail("AUDIT_HASH_MISMATCH", body["seq"])
        jwk = key_of.get(current_kid)
        if jwk is None:
            return fail("AUDIT_TRUST_INVALID", body["seq"])
        if not verify_entry(e, jwk):
            return fail("AUDIT_SIGNATURE_INVALID", body["seq"])
        # signer rotation admission
        if body["kind"] == "key.rotated" and body.get("key_id") in key_of:
            current_kid = body["key_id"]
        prev = e["hash"]
        expect_seq += 1
        last = e

    head = page["head"]
    head_jwk = key_of.get(head["signer_kid"])
    if head_jwk is None:
        return fail("AUDIT_TRUST_INVALID", head["seq"])
    if not verify_head(head, trust["tenant_id"], head_jwk):
        return fail("AUDIT_SIGNATURE_INVALID", head["seq"])
    if last is None:
        cp = trust.get("checkpoint")
        if cp is None or cp["hash"] != head["hash"] or cp["body"]["seq"] != head["seq"]:
            return fail("AUDIT_INCOMPLETE", head["seq"])
    else:
        if head["seq"] != last["body"]["seq"] or head["hash"] != last["hash"]:
            return fail("AUDIT_INCOMPLETE", str(int(last["body"]["seq"]) + 1))
    pinned = trust.get("pinned_head")
    if pinned is not None:
        if (pinned["seq"] != head["seq"] or pinned["hash"] != head["hash"]
                or pinned["signer_kid"] != head["signer_kid"]):
            return fail("AUDIT_INCOMPLETE", head["seq"])
    return {"valid": True, "entries": len(entries), "head": head["hash"]}


def valid_audit_page(page) -> bool:
    if not isinstance(page, dict):
        return False
    if page.get("v") != 1 or not isinstance(page.get("entries"), list):
        return False
    if not is_counter(page.get("next_after")) or not isinstance(page.get("has_more"), bool):
        return False
    h = page.get("head")
    if not (isinstance(h, dict) and is_counter(h.get("seq")) and is_hash(h.get("hash"))
            and is_id(h.get("signer_kid"), "lky") and isinstance(h.get("signature"), str)):
        return False
    for e in page["entries"]:
        if not (isinstance(e, dict) and isinstance(e.get("body"), dict)
                and is_hash(e.get("hash")) and isinstance(e.get("signature"), str)):
            return False
    return True


def valid_audit_trust(t) -> bool:
    if not isinstance(t, dict):
        return False
    if t.get("v") != 1 or not is_id(t.get("tenant_id"), "ltn") or not is_id(t.get("initial_kid"), "lky"):
        return False
    keys = t.get("keys")
    if not isinstance(keys, list):
        return False
    prev = ""
    for k in keys:
        if not (isinstance(k, dict) and is_id(k.get("kid"), "lky")):
            return False
        j = k.get("jwk")
        if not (isinstance(j, dict) and j.get("kty") == "OKP" and j.get("crv") == "Ed25519"
                and isinstance(j.get("x"), str) and len(b64u_decode(j["x"]) or b"") == 32):
            return False
        if prev and prev >= k["kid"]:
            return False
        prev = k["kid"]
    # pinned_head / checkpoint must be present (null or a valid member);
    # absent members are rejected like the TS parser's `!== null` check
    if "pinned_head" not in t:
        return False
    h = t["pinned_head"]
    if h is not None and not (
        isinstance(h, dict) and is_counter(h.get("seq")) and is_hash(h.get("hash"))
        and is_id(h.get("signer_kid"), "lky") and isinstance(h.get("signature"), str)
    ):
        return False
    if "checkpoint" not in t:
        return False
    cp = t["checkpoint"]
    if cp is not None and not (
        isinstance(cp, dict) and isinstance(cp.get("body"), dict)
        and is_hash(cp.get("hash")) and isinstance(cp.get("signature"), str)
    ):
        return False
    return True

File: src/core/audit.ts
// Signed hash-chained audit per §11.1.
// entry.hash = D("LEXSCOPE-AUDIT/1", body)
// signature   = Ed25519(UTF8("LEXSCOPE-AUDIT-SIGN/1") || 0x00 || U(hash))
// head sig    = Ed25519(UTF8("LEXSCOPE-HEAD/1") || 0x00 || J({tenant_id,seq,hash,signer_kid}))

import { KeyObject } from "node:crypto";
import { domainHash, unhex } from "./hash.ts";
import { jcs } from "./jcs.ts";
import { b64uEncode, b64uDecode } from "./b64.ts";
import { ed25519Sign, ed25519Verify, type PublicJwk } from "./jwt.ts";
import type { ErrorCode } from "./errors.ts";
import { isId, isCounter, idValidators } from "./ids.ts";
import { utf16Compare } from "./jcs.ts";

export const EVENT_KINDS = new Set([
  "deployment.installed", "token.minted", "token.expired", "task.expired",
  "nonce.issued", "nonce.expired", "call.admitted", "call.succeeded",
  "call.failed", "call.unknown", "call.output_blocked", "result.read",
  "access.denied", "inspect.read", "revocation.applied", "policy.applied",
  "key.rotated", "key.retired", "audit.read", "nonce.read", "control.replayed",
  "storage.migrated",
]);
export type EventKind = string;

export interface AuditBody {
  v: 1;
  schema: "lexscope.audit/1";
  tenant_id: string;
  seq: string;
  event_id: string;
  ts: number;
  prev: string;
  signer_kid: string;
  kind: EventKind;
  sub: string | null;
  task_id: string | null;
  token_id: string | null;
  call_id: string | null;
  request_id: string | null;
  control_kid: string | null;
  policy_hash: string | null;
  code: ErrorCode | "OK";
  call_hash: string | null;
  scope_hash: string | null;
  nonce_id: string | null;
  target: { kind: string; id: string } | null;
  revision: string | null;
  key_id: string | null;
}

export interface AuditEntry {
  body: AuditBody;
  hash: string;
  signature: string;
}

export interface AuditHead {
  seq: string;
  hash: string;
  signer_kid: string;
  signature: string;
}

export interface AuditPage {
  v: 1;
  entries: AuditEntry[];
  next_after: string;
  head: AuditHead;
  has_more: boolean;
}

export interface AuditTrust {
  v: 1;
  tenant_id: string;
  initial_kid: string;
  keys: { kid: string; jwk: PublicJwk }[];
  pinned_head: AuditHead | null;
  checkpoint: AuditEntry | null;
}

export type AuditVerifyResult =
  | { valid: true; entries: number; head: string }
  | {
      valid: false;
      code: "AUDIT_HASH_MISMATCH" | "AUDIT_INCOMPLETE" | "AUDIT_SIGNATURE_INVALID" | "AUDIT_SEQUENCE_INVALID" | "AUDIT_TRUST_INVALID";
      seq: string;
    };

export const GENESIS_PREV = "0".repeat(64);

export function makeAuditBody(partial: Partial<AuditBody> & Pick<AuditBody, "tenant_id" | "seq" | "event_id" | "ts" | "prev" | "signer_kid" | "kind">): AuditBody {
  return {
    v: 1,
    schema: "lexscope.audit/1",
    sub: null,
    task_id: null,
    token_id: null,
    call_id: null,
    request_id: null,
    control_kid: null,
    policy_hash: null,
    code: "OK",
    call_hash: null,
    scope_hash: null,
    nonce_id: null,
    target: null,
    revision: null,
    key_id: null,
    ...partial,
  };
}

export function entryHash(body: AuditBody): string {
  return domainHash("LEXSCOPE-AUDIT/1", body);
}

export function signEntry(body: AuditBody, key: KeyObject): AuditEntry {
  const hash = entryHash(body);
  const sig = ed25519Sign(Buffer.concat([Buffer.from("LEXSCOPE-AUDIT-SIGN/1", "ascii"), Buffer.from([0]), unhex(hash)!]), key);
  return { body, hash, signature: b64uEncode(sig) };
}

export function verifyEntry(entry: AuditEntry, jwk: PublicJwk): boolean {
  const sig = b64uDecode(entry.signature);
  const digest = unhex(entry.hash);
  if (sig === null || sig.length !== 64 || digest === null) return false;
  return ed25519Verify(
    Buffer.concat([Buffer.from("LEXSCOPE-AUDIT-SIGN/1", "ascii"), Buffer.from([0]), digest]),
    sig,
    jwk,
  );
}

export function signHead(tenantId: string, seq: string, hash: string, signerKid: string, key: KeyObject): AuditHead {
  const msg = Buffer.concat([
    Buffer.from("LEXSCOPE-HEAD/1", "ascii"),
    Buffer.from([0]),
    jcs({ tenant_id: tenantId, seq, hash, signer_kid: signerKid }),
  ]);
  return { seq, hash, signer_kid: signerKid, signature: b64uEncode(ed25519Sign(msg, key)) };
}

export function verifyHead(head: AuditHead, tenantId: string, jwk: PublicJwk): boolean {
  const sig = b64uDecode(head.signature);
  if (sig === null || sig.length !== 64) return false;
  const msg = Buffer.concat([
    Buffer.from("LEXSCOPE-HEAD/1", "ascii"),
    Buffer.from([0]),
    jcs({ tenant_id: tenantId, seq: head.seq, hash: head.hash, signer_kid: head.signer_kid }),
  ]);
  return ed25519Verify(msg, sig, jwk);
}

// ---------- structural validation of exported artifacts ----------

function isAuditBody(v: unknown): v is AuditBody {
  if (v === null || typeof v !== "object" || Array.isArray(v)) return false;
  const b = v as Record<string, unknown>;
  return (
    b.v === 1 &&
    b.schema === "lexscope.audit/1" &&
    isId(b.tenant_id, "ltn") &&
    isCounter(b.seq) &&
    isId(b.event_id, "lev") &&
    typeof b.ts === "number" &&
    typeof b.prev === "string" &&
    isId(b.signer_kid, "lky") &&
    typeof b.kind === "string" &&
    EVENT_KINDS.has(b.kind as string)
  );
}

function isAuditEntry(v: unknown): v is AuditEntry {
  if (v === null || typeof v !== "object") return false;
  const e = v as Record<string, unknown>;
  return isAuditBody(e.body) && typeof e.hash === "string" && /^[0-9a-f]{64}$/.test(e.hash) && typeof e.signature === "string";
}

function isAuditHead(v: unknown): v is AuditHead {
  if (v === null || typeof v !== "object") return false;
  const h = v as Record<string, unknown>;
  return isCounter(h.seq) && typeof h.hash === "string" && /^[0-9a-f]{64}$/.test(h.hash) && isId(h.signer_kid, "lky") && typeof h.signature === "string";
}

export function parseAuditTrust(v: unknown): AuditTrust | null {
  if (v === null || typeof v !== "object") return null;
  const t = v as Record<string, unknown>;
  if (t.v !== 1 || !isId(t.tenant_id, "ltn") || !isId(t.initial_kid, "lky") || !Array.isArray(t.keys)) return null;
  const keys = t.keys as unknown[];
  let prev = "";
  for (const k of keys) {
    if (k === null || typeof k !== "object") return null;
    const kk = k as Record<string, unknown>;
    if (!isId(kk.kid, "lky")) return null;
    const j = kk.jwk as Record<string, unknown> | undefined;
    if (!j || j.kty !== "OKP" || j.crv !== "Ed25519" || typeof j.x !== "string" || b64uDecode(j.x)?.length !== 32) return null;
    if (prev !== "" && prev >= (kk.kid as string)) return null;
    prev = kk.kid as string;
  }
  if (t.pinned_head !== null && !isAuditHead(t.pinned_head)) return null;
  if (t.checkpoint !== null && !isAuditEntry(t.checkpoint)) return null;
  return t as unknown as AuditTrust;
}

export function parseAuditPage(v: unknown): AuditPage | null {
  if (v === null || typeof v !== "object") return null;
  const p = v as Record<string, unknown>;
  if (p.v !== 1 || !Array.isArray(p.entries) || !isCounter(p.next_after) || !isAuditHead(p.head) || typeof p.has_more !== "boolean") return null;
  for (const e of p.entries) if (!isAuditEntry(e)) return null;
  return p as unknown as AuditPage;
}

// ---------- offline verification (§11.1 tail) ----------

// Verifies a complete AuditPage (as produced by `audit export`) against an
// AuditTrust. Completeness requires reaching the signed head AND matching any
// independently pinned head.
export function verifyAudit(page: AuditPage, trust: AuditTrust): AuditVerifyResult {
  if (trust.tenant_id !== "" ) {
    // tenant must match the entries' tenant; head sig checked below
  }
  const keyOf = (kid: string) => trust.keys.find((k) => k.kid === kid)?.jwk ?? null;
  const fail = (code: "AUDIT_HASH_MISMATCH" | "AUDIT_INCOMPLETE" | "AUDIT_SIGNATURE_INVALID" | "AUDIT_SEQUENCE_INVALID" | "AUDIT_TRUST_INVALID", seq: string): AuditVerifyResult => ({ valid: false, code, seq });

  // Entry sequence start: genesis unless a trusted checkpoint anchors it.
  let expectSeq: bigint;
  let expectPrev: string;
  let signer: string;
  if (trust.checkpoint !== null) {
    const cp = trust.checkpoint;
    if (!isAuditBody(cp.body) || entryHash(cp.body) !== cp.hash) return fail("AUDIT_HASH_MISMATCH", String(cp.body?.seq ?? "0"));
    const cpJwk = keyOf(cp.body.signer_kid);
    if (!cpJwk) return fail("AUDIT_TRUST_INVALID", cp.body.seq);
    if (!verifyEntry(cp, cpJwk)) return fail("AUDIT_SIGNATURE_INVALID", cp.body.seq);
    if (cp.body.tenant_id !== trust.tenant_id) return fail("AUDIT_TRUST_INVALID", cp.body.seq);
    expectSeq = BigInt(cp.body.seq) + 1n;
    expectPrev = cp.hash;
    signer = cp.body.signer_kid;
  } else {
    expectSeq = 1n;
    expectPrev = GENESIS_PREV;
    signer = trust.initial_kid;
  }
  if (!keyOf(signer)) return fail("AUDIT_TRUST_INVALID", "0");

  let last: AuditEntry | null = null;
  for (const e of page.entries) {
    const b = e.body;
    const seqStr = String(b.seq);
    if (!isCounter(b.seq) || BigInt(b.seq) !== expectSeq) return fail("AUDIT_SEQUENCE_INVALID", seqStr);
    if (b.tenant_id !== trust.tenant_id || b.schema !== "lexscope.audit/1" || b.v !== 1) return fail("AUDIT_TRUST_INVALID", seqStr);
    if (b.prev !== expectPrev) return fail("AUDIT_HASH_MISMATCH", seqStr);
    if (entryHash(b) !== e.hash) return fail("AUDIT_HASH_MISMATCH", seqStr);
    // signer timeline: current signer, or a rotation boundary event
    if (b.signer_kid !== signer) {
      if (b.kind === "key.rotated" && keyOf(b.signer_kid) && b.key_id === b.signer_kid) {
        signer = b.signer_kid;
      } else return fail("AUDIT_TRUST_INVALID", seqStr);
    }
    const jwk = keyOf(b.signer_kid);
    if (!jwk) return fail("AUDIT_TRUST_INVALID", seqStr);
    if (!verifyEntry(e, jwk)) return fail("AUDIT_SIGNATURE_INVALID", seqStr);
    expectSeq += 1n;
    expectPrev = e.hash;
    last = e;
  }

  // The signed page head must be exactly the last verified entry.
  const head = page.head;
  const headJwk = keyOf(head.signer_kid);
  if (!headJwk) return fail("AUDIT_TRUST_INVALID", head.seq);
  if (!verifyHead(head, trust.tenant_id, headJwk)) return fail("AUDIT_SIGNATURE_INVALID", head.seq);
  if (last === null) {
    // empty retained history verifies only against a matching signed checkpoint/head
    if (trust.checkpoint === null || trust.checkpoint.hash !== head.hash || trust.checkpoint.body.seq !== head.seq)
      return fail("AUDIT_INCOMPLETE", head.seq);
  } else {
    if (head.seq !== last.body.seq || head.hash !== last.hash) return fail("AUDIT_INCOMPLETE", String(BigInt(last.body.seq) + 1n));
  }
  if (trust.pinned_head !== null) {
    if (trust.pinned_head.seq !== head.seq || trust.pinned_head.hash !== head.hash || trust.pinned_head.signer_kid !== head.signer_kid)
      return fail("AUDIT_INCOMPLETE", head.seq);
  }
  return { valid: true, entries: page.entries.length, head: head.hash };
}
